rclcpp/include/rclcpp/qos_event.hpp

```diff
@@ -0,0 +1,164 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__QOS_EVENT_HPP_
+#define RCLCPP__QOS_EVENT_HPP_
+
+#include <functional>
+
+#include "rcl/error_handling.h"
+
+#include "rcutils/logging_macros.h"
+
+#include "rclcpp/waitable.hpp"
+#include "rclcpp/exceptions.hpp"
+#include "rclcpp/function_traits.hpp"
+
+
+namespace rclcpp
+{
+
+struct QOSDeadlineEventInfo
+{
+  enum DeadlineEventType
+  {
+    DEADLINE_MISSED
+  };
+
+  DeadlineEventType event_type;
+  void * other_info;
+};
+
+struct QOSLivelinessEventInfo
+{
+  enum LivelinessEventType
+  {
+    LIVELINESS_CHANGED
+  };
+
+  LivelinessEventType event_type;
+  void * other_info;
+};
+
+struct QOSLifespanEventInfo
+{
+  enum LifespanEventType
+  {
+    LIFESPAN_EXPIRED
+  };
+
+  LifespanEventType event_type;
+  void * other_info;
+};
+
+
+using QOSDeadlineEventCallbackType = std::function<void (QOSDeadlineEventInfo &)>;
+using QOSLivelinessEventCallbackType = std::function<void (QOSLivelinessEventInfo &)>;
+using QOSLifespanEventCallbackType = std::function<void (QOSLifespanEventInfo &)>;
+
+
+// struct QOSDeadlineEventConfig {
+//   using CallbackT = QOSDeadlineEventCallbackType;
+//   using EventInfoT = QOSDeadlineEventInfo;
+// };
+
+// struct QOSLivelinessEventConfig {
+//   using CallbackT = QOSLivelinessEventCallbackType;
+//   using EventInfoT = QOSLivelinessEventInfo;
+// };
+
+// struct QOSLifespanEventConfig {
+//   using CallbackT = QOSLifespanEventCallbackType;
+//   using EventInfoT = QOSLifespanEventInfo;
+// };
+
+
+class QOSEventBase : public Waitable
+{
+public:
+  /// Get the number of ready events
+  size_t
+  get_number_of_ready_events() override;
+
+  /// Add the Waitable to a wait set.
+  bool
+  add_to_wait_set(rcl_wait_set_t * wait_set) override;
+
+  /// Check if the Waitable is ready.
+  bool
+  is_ready(rcl_wait_set_t * wait_set) override;
+
+protected:
+  rcl_event_t event_handle_;
+  size_t wait_set_event_index_;
+};
+
+
+template<typename EventCallbackT>
+class QOSEvent : public QOSEventBase
+{
+public:
+  template<typename InitFuncT, typename HandleT, typename OptionsT, typename EventTypeEnum>
```

Comment:
If this has reached the complexity of needing to template initialization functions and types that relate to the specific callback then it seems like we should be using polymorphism to extend the base class instead of templating further.

Reply:
I don't think defining _n_ number of derived classes with a bunch of member functions here can scale with the _n_ number of simple QoS status struct types defined in `rmw/types.h`

Reply:
I don't think it needs a bunch of member functions. Each derived class would probably only need a constructor or to implement a virtual initialize function that the constructor calls with the `event_handle_`, which you already need to create for every status type. I think it's fine to still keep the base class templated so that the destructor and execute functions can be reused, but templating an initialize functions is where it starts to feel like a hack around inheritance to me. Yes it's a couple extra lines of code for every status, but in my opinion it's easier to understand code.

Reply:
Though I generally prefer polymorphism I don't know if it's the best path forward because there's no real commonality between the base events defined in rmw (e.g. a base class doesn't make sense). We're lucky that 3/4 have the same fields, but that won't be the case if we look to QoS types beyond DDS.

Reply:
The commonality is that these are all interfaces for handling events. All events come out of the same take function in RMW when an event is ready and they all pass that along to some user provided callback function. The divergence is in the type for the event. If the code is exactly the same and the only difference is in the type then templates are definitely the right way to go. But in this case it's actually using another template parameter to change out the code that's being executed by providing different initialization functions based on the type. To me that seems like a much better case for polymorphism. 

From the offline conversation, I still feel like the combination of polymorphism and templating are the cleaner approach, but I'm willing to consider it a stylistic difference because I don't know that there is a strong technical argument for going one way or the other.

Reply:
Calling initialization functions that have different number of parameters can still be handled by variadic templates, so I think this is still a stylistic difference.